src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,123 @@
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer key is not a subset of inner key
+                if !group_key
+                    .iter()
+                    .all(|e| matches!(e, MirScalarExpr::Column(_)) && inner_group_key.contains(e))
+                {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Replace inner reduce with map + project (no grouping)
+                    let mut outputs = vec![];
+                    let mut scalars = vec![];
+                    for e in inner_group_key {
+                        if let MirScalarExpr::Column(i) = e {
+                            outputs.push(*i);
+                        } else {
+                            scalars.push(e.clone());
+                        }
+                    }
+
+                    let arity = inner_input.arity();
+                    for i in 0..scalars.len() {
+                        outputs.push(arity + i);
+                    }
+
+                    **input = inner_input.take_dangerous().map(scalars).project(outputs);
+                } else if aggregates.len() == 1 && inner_aggregates.len() == 1 {
+                    // TODO: This can be more general than just len() == 1
+                    // Drop inner reduce and rewrite outer reduce if possible
+                    if let AggregateExpr {
+                        func: AggregateFunc::Count,
+                        expr,
+                        distinct: _,
+                    } = inner_aggregates.first().unwrap()
+                    {
+                        if inner_group_key.contains(expr) {
+                            return;
+                        }
+
+                        if let AggregateExpr {
+                            func: AggregateFunc::SumInt64,
+                            expr: MirScalarExpr::Column(i),
+                            distinct: _,
+                        } = aggregates.first().unwrap()
+                        {
+                            if *i != inner_group_key.len() {
+                                return;
+                            }
+
+                            aggregates[0] = inner_aggregates[0].clone();
```

Comment:
This messes up again with column offsets.

```
create table a(a int, b int, c int);
select a, sum(co) from (select b, a, count(b) as co from a group by b, a) group by a;
```

Given that example, the current code will result in the plan for:

```
select a, count(a) from a group by a;
```

instead of the plan of the following equivalent query:

```
select a, count(b) from a group by a;
```

due to the reshuffling of the columns the inner reduction does.

However, as mentioned above, I'd leave this out of the scope of the first patch.

Reply:
> However, as mentioned above, I'd leave this out of the scope of the first patch.

I agree and removed the code for now. 👍🏼